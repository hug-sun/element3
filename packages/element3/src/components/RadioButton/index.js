import ElRadioButton from './src/RadioButton.vue'

/* istanbul ignore next */
ElRadioButton.install = function (app) {
  app.component(ElRadioButton.name, ElRadioButton)
}

export { ElRadioButton }
