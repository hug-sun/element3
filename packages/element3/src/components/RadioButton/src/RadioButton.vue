<template>
  <label
    class="el-radio-button"
    :class="classes"
    role="radio"
    :aria-checked="isChecked"
    :aria-disabled="isDisabled"
    :tabindex="tabIndex"
    @keydown.space.stop.prevent="value = isDisabled ? value : label"
  >
    <input
      class="el-radio-button__orig-radio"
      type="radio"
      @change="handleChange"
      :value="label"
      :name="name"
      v-model="value"
      :disabled="isDisabled"
      @focus="focus = true"
      @blur="focus = false"
      tabindex="-1"
    />
    <span
      class="el-radio-button__inner"
      :style="isChecked ? style : null"
      @keydown.stop
    >
      <slot>{{ label }}</slot>
    </span>
  </label>
</template>
<script>
import { toRefs, ref, inject, computed, watch, getCurrentInstance } from 'vue'
import { props } from './props'
import { useGlobalOptions } from '../../../composables/globalConfig'

export default {
  name: 'ElRadioButton',

  componentName: 'ElRadioButton',

  props,

  setup(props) {
    const { radioGroup } = useCheckGroup()
    const { label, disabled } = toRefs(props)
    const focus = ref(false)

    let value = useModel(radioGroup)
    const handleChange = useChange(radioGroup, value)
    const { style, size, isDisabled } = useStyle({ disabled, radioGroup })
    const { classes, isChecked, tabIndex } = useClasses({
      size,
      isDisabled,
      value,
      label,
      focus,
      radioGroup
    })

    return {
      value,
      handleChange,
      isDisabled,
      tabIndex,
      classes,
      isChecked,
      style
    }
  }
}

const useCheckGroup = () => {
  const { parent } = getCurrentInstance()
  const isGroup = parent.type.name === 'ElRadioGroup'
  const radioGroup = isGroup ? parent : null
  return { isGroup, radioGroup }
}

const useModel = (radioGroup) => {
  const value = ref(radioGroup?.props.modelValue)

  watch(
    () => radioGroup?.props.modelValue,
    (val) => {
      value.value = val
    }
  )

  return value
}

const useChange = (radioGroup, value) => {
  const handleChange = () => {
    radioGroup?.emit('update:modelValue', value.value)
    radioGroup?.emit('change', value.value)
  }

  return handleChange
}

const useStyle = ({ disabled, radioGroup }) => {
  const globalConfig = useGlobalOptions()

  const isDisabled = computed(() => {
    const elForm = inject('elForm', {})

    return disabled?.value || radioGroup?.props.disabled || elForm.disabled
  })

  const size = computed(() => {
    const elFormItem = inject('elFormItem', {})

    return (
      radioGroup?.props.size || elFormItem?.elFormItemSize || globalConfig.size
    )
  })

  const style = computed(() => {
    return {
      backgroundColor: radioGroup?.props.fill || '',
      borderColor: radioGroup?.props.fill || '',
      boxShadow: radioGroup?.props.fill
        ? `-1px 0 0 0 ${radioGroup?.props.fill}`
        : '',
      color: radioGroup?.props.textColor || ''
    }
  })

  return {
    isDisabled,
    size,
    style
  }
}

const useClasses = ({ size, isDisabled, focus, value, label, radioGroup }) => {
  const isChecked = computed(() => {
    return value.value === label.value
  })

  const tabIndex = computed(() => {
    return isDisabled.value || (radioGroup && value.value !== label.value)
      ? -1
      : 0
  })

  const classes = computed(() => {
    return [
      size.value ? 'el-radio-button--' + size.value : '',
      { 'is-active': isChecked.value },
      { 'is-disabled': isDisabled.value },
      { 'is-focus': focus.value }
    ]
  })

  return {
    isChecked,
    classes,
    tabIndex
  }
}
</script>
