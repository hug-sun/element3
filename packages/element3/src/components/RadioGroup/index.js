import ElRadioGroup from './src/RadioGroup'

ElRadioGroup.install = function (app) {
  app.component(ElRadioGroup.name, ElRadioGroup)
}

export { ElRadioGroup }
