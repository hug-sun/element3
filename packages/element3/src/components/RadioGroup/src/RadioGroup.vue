<template>
  <div class="el-radio-group" role="radiogroup">
    <slot></slot>
  </div>
</template>

<script lang="ts">
import { props } from './props'
import { defineComponent, inject, computed, toRefs, watch } from 'vue'
import { useGlobalOptions } from '../../../composables/globalConfig'
import { useEmitter } from '../../../composables/emitter'

export default defineComponent({
  name: 'ElRadioGroup',

  props,

  emits: ['update:modelValue', 'change'],

  setup(props) {
    const { size, modelValue } = toRefs(props)
    const globalConfig = useGlobalOptions()
    const elFormItem = inject('elFormItem', {})
    const { dispatch } = useEmitter()

    watch(modelValue, (v) => {
      dispatch('el.form.change', v)
    })

    const radioGroupSize = useRadioGroupSize({ size, elFormItem, globalConfig })

    return {
      radioGroupSize
    }
  }
})

const useRadioGroupSize = ({ size, elFormItem, globalConfig }) => {
  return computed(() => {
    return (
      size?.value || (elFormItem as any)?.elFormItemSize || globalConfig.size
    )
  })
}
</script>
