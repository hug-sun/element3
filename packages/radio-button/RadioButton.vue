<template>
  <label
    class="el-radio-button"
    :class="[
      size ? 'el-radio-button--' + size : '',
      { 'is-active': value === label },
      { 'is-disabled': isDisabled },
      { 'is-focus': focus }
    ]"
    role="radio"
    :aria-checked="value === label"
    :aria-disabled="isDisabled"
    :tabindex="tabIndex"
    @keydown.space.stop.prevent="value = isDisabled ? value : label"
  >
    <input
      class="el-radio-button__orig-radio"
      :value="label"
      type="radio"
      v-model="value"
      :name="name"
      @change="handleChange"
      :disabled="isDisabled"
      tabindex="-1"
      @focus="focus = true"
      @blur="focus = false"
    />
    <span
      class="el-radio-button__inner"
      :style="value === label ? activeStyle : null"
      @keydown.stop
    >
      <slot></slot>
      <template v-if="!$slots.default">{{ label }}</template>
    </span>
  </label>
</template>
<script>
import {
  toRefs,
  ref,
  inject,
  getCurrentInstance,
  computed,
  nextTick
} from 'vue'

export default {
  name: 'ElRadioButton',

  componentName: 'ElRadioButton',

  props: {
    label: [String, Number, Symbol, Boolean],
    disabled: Boolean,
    name: String
  },

  setup(props) {
    const { label, disabled } = toRefs(props)
    const radio = ref()
    const focus = ref(false)

    const { elForm, elFormItem } = useInject()

    const { radioGroup } = useCheckGroup()

    const { value, handleChange } = useModel({ radioGroup, label, radio })

    const { size, isDisabled, tabIndex, activeStyle } = useStyle({
      radioGroup,
      disabled,
      value,
      label,
      elForm,
      elFormItem
    })

    return {
      radio,
      focus,
      value,
      size,
      isDisabled,
      tabIndex,
      activeStyle,
      handleChange
    }
  }
}

function useInject() {
  const elForm = inject('elForm', {})
  const elFormItem = inject('elFormItem', {})
  return {
    elForm,
    elFormItem
  }
}

function useCheckGroup() {
  let { parent } = getCurrentInstance()
  while (parent) {
    if (parent.type.name !== 'ElRadioGroup') {
      parent = parent.parent
    } else {
      return {
        isGroup: true,
        radioGroup: parent
      }
    }
  }
  console.warn(
    'ElementUI Warn: `<radio-button></radio-button>` must be use with <radio-group></radio-group>'
  )
  return {
    isGroup: false,
    radioGroup: null
  }
}

function useModel({ radioGroup }) {
  const value = computed({
    get() {
      return radioGroup.props.modelValue
    },
    set(val) {
      radioGroup.emit('update:modelValue', val)
      // radio.checked = value.value === label.value;
    }
  })

  async function handleChange() {
    await nextTick()
    radioGroup.emit('change', value.value)
  }
  return { value, handleChange }
}

function useStyle({ radioGroup, disabled, value, label, elForm, elFormItem }) {
  const { ctx } = getCurrentInstance()
  const elFormDisable = elForm.disabled

  const size = computed(() => {
    const temRadioSize = elFormItem.elFormItemSize || (ctx.$ELEMENT || {}).size
    return radioGroup.ctx.radioGroupSize || temRadioSize
  })
  const isDisabled = computed(() => {
    return radioGroup.props.disabled || disabled.value || elFormDisable
  })
  const tabIndex = computed(() => {
    return isDisabled.value || (radioGroup && value.value !== label.value)
      ? -1
      : 0
  })

  const activeStyle = computed(() => {
    return {
      backgroundColor: radioGroup.props.fill || '',
      borderColor: radioGroup.props.fill || '',
      boxShadow: radioGroup.props.fill
        ? `-1px 0 0 0 ${radioGroup.props.fill}`
        : '',
      color: radioGroup.props.textColor || ''
    }
  })

  return {
    size,
    isDisabled,
    tabIndex,
    activeStyle
  }
}
</script>
