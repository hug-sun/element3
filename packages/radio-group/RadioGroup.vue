<template @change="test">
  <component
    :is="elTag"
    class="el-radio-group"
    role="radiogroup"
    @keydown="handleKeydown"
  >
    <slot></slot>
  </component>
</template>
<script>
import { useEmitter } from '../../src/use/emitter'
import {
  getCurrentInstance,
  onMounted,
  watch,
  toRefs,
  computed,
  inject
} from 'vue'
const keyCode = Object.freeze({
  LEFT: 37,
  UP: 38,
  RIGHT: 39,
  DOWN: 40
})

export default {
  name: 'ElRadioGroup',

  componentName: 'ElRadioGroup',

  props: {
    modelValue: [String, Number, Symbol, Boolean],
    size: String,
    fill: String,
    textColor: String,
    disabled: Boolean
  },

  emits: ['update:modelValue', 'change'],

  setup(props) {
    const { modelValue = {}, size = {} } = toRefs(props)
    const { dispatch } = useEmitter()
    const { elFormItem } = useInject()

    watch(modelValue, (v) => {
      dispatch('el.form.change', v)
    })

    const { handleKeydown } = useKeyDown()

    const { elTag, radioGroupSize } = useStyle({ elFormItem, size })

    return {
      elTag,
      radioGroupSize,
      handleKeydown
    }
  }
}

function useInject() {
  const elFormItem = inject('elFormItem', {})
  return {
    elFormItem
  }
}

function useKeyDown() {
  const { ctx } = getCurrentInstance()
  onMounted(() => {
    const radios = ctx.$el.querySelectorAll('[type=radio]')
    const firstLabel = ctx.$el.querySelectorAll('[role=radio]')[0]
    if (![].some.call(radios, (radio) => radio.checked) && firstLabel) {
      firstLabel.tabIndex = 0
    }
  })

  function handleKeydown(e) {
    const target = e.target
    const className =
      target.nodeName === 'INPUT' ? '[type=radio]' : '[role=radio]'
    const radios = ctx.$el.querySelectorAll(className)
    const length = radios.length
    const index = [].indexOf.call(radios, target)
    const roleRadios = ctx.$el.querySelectorAll('[role=radio]')
    switch (e.keyCode) {
      case keyCode.LEFT:
      case keyCode.UP:
        e.stopPropagation()
        e.preventDefault()
        if (index === 0) {
          roleRadios[length - 1].click()
          roleRadios[length - 1].focus()
        } else {
          roleRadios[index - 1].click()
          roleRadios[index - 1].focus()
        }
        break
      case keyCode.RIGHT:
      case keyCode.DOWN:
        if (index === length - 1) {
          e.stopPropagation()
          e.preventDefault()
          roleRadios[0].click()
          roleRadios[0].focus()
        } else {
          roleRadios[index + 1].click()
          roleRadios[index + 1].focus()
        }
        break
      default:
        break
    }
  }

  return { handleKeydown }
}

function useStyle({ elFormItem, size }) {
  const { ctx, vnode } = getCurrentInstance()

  const radioGroupSize = computed(() => {
    return size.value || elFormItem.elFormItemSize || (ctx.$ELEMENT || {}).size
  })

  const elTag = computed(() => {
    return (vnode.data || {}).tag || 'div'
  })

  return { radioGroupSize, elTag }
}
</script>
