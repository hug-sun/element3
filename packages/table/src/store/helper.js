import Store from './index'
import { debounce } from 'lodash'

export function createStore(table, initialState = {}) {
  if (!table) {
    throw new Error('Table is required.')
  }

  const store = Store()
  // fix https://github.com/ElemeFE/element/issues/14075
  // related pr https://github.com/ElemeFE/element/pull/14146
  store.toggleAllSelection = debounce(store._toggleAllSelection, 10)
  Object.keys(initialState).forEach((key) => {
    store.states[key].value = initialState[key]
  })
  return store
}

export function mapStates(mapper) {
  const res = {}
  Object.keys(mapper).forEach((key) => {
    const value = mapper[key]
    let fn
    if (typeof value === 'string') {
      fn = function () {
        return this.store.states[value]
      }
    } else if (typeof value === 'function') {
      fn = function () {
        return value.call(this, this.store.states)
      }
    } else {
      console.error('invalid value type')
    }
    if (fn) {
      res[key] = fn
    }
  })
  return res
}
