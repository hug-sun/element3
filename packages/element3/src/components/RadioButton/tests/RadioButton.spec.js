import RadioButton from '../src/RadioButton.vue'
import RadioGroup from '../../RadioGroup/src/RadioGroup.vue'
import { mount } from '@vue/test-utils'
import { expectHaveClass } from '../../../../tests/helper'
import { ref, h, nextTick, reactive } from 'vue'

const content = ['上海', '北京']

describe('radioButton.vue', () => {
  it('snapshot', () => {
    const wrapper = mount(RadioButton)
    expect(wrapper.element).toMatchSnapshot()
  })
  it('should show content', () => {
    const wrapper = mount(RadioGroup, {
      props: {
        size: 'mini'
      },
      slots: {
        default: content.map((label) => h(RadioButton, { label }))
      }
    })

    wrapper.findAllComponents({ name: 'ElRadioButton' }).forEach((item, i) => {
      expect(item.text()).toContain(content[i])
    })
  })

  it('size', async () => {
    const wrapper = mount(RadioGroup, {
      slots: {
        default: content.map((label) => h(RadioButton, { label }))
      }
    })

    await wrapper.setProps({ size: 'mini' })
    expectHaveClass(
      wrapper.findComponent({ name: 'ElRadioButton' }),
      `el-radio-button--mini`
    )

    await wrapper.setProps({ size: 'small' })

    expectHaveClass(
      wrapper.findComponent({ name: 'ElRadioButton' }),
      `el-radio-button--small`
    )

    await wrapper.setProps({ size: 'medium' })

    expectHaveClass(
      wrapper.findComponent({ name: 'ElRadioButton' }),
      `el-radio-button--medium`
    )
  })

  it('modelValue', async () => {
    const modelValue = ref('')
    const wrapper = mount(RadioGroup, {
      props: {
        modelValue,
        'onUpdate:modelValue'(v) {
          modelValue.value = v
        }
      },
      slots: {
        default: content.map((label) => h(RadioButton, { label }))
      }
    })

    await wrapper
      .findAllComponents({ name: 'ElRadioButton' })[0]
      .trigger('click')
    expect(modelValue.value).toStrictEqual('上海')

    await wrapper
      .findAllComponents({ name: 'ElRadioButton' })[1]
      .trigger('click')
    expect(modelValue.value).toStrictEqual('北京')
  })

  it('disabled', async () => {
    const modelValue = ref('')
    const wrapper = mount(RadioGroup, {
      props: {
        modelValue,
        'onUpdate:modelValue'(v) {
          modelValue.value = v
        }
      },
      slots: {
        default: content.map((label) => h(RadioButton, { label }))
      }
    })

    await wrapper.setProps({ disabled: true })
    expectHaveClass(
      wrapper.findComponent({ name: 'ElRadioButton' }),
      'is-disabled'
    )

    await wrapper
      .findAllComponents({ name: 'ElRadioButton' })[1]
      .trigger('click')
    expect(modelValue.value).toStrictEqual('')
  })

  it('by elForm.disable', () => {
    const wrapper = mount(RadioButton, {
      global: {
        provide: {
          elForm: reactive({
            disabled: true
          })
        }
      }
    })
    expect(wrapper.classes()).toContain(`is-disabled`)
  })

  it('is-active', async () => {
    const modelValue = ref('')
    const wrapper = mount(RadioGroup, {
      props: {
        modelValue,
        'onUpdate:modelValue'(v) {
          modelValue.value = v
        }
      },
      slots: {
        default: content.map((label) => h(RadioButton, { label }))
      }
    })

    await wrapper
      .findAllComponents({ name: 'ElRadioButton' })[0]
      .trigger('click')

    expectHaveClass(
      wrapper.findAllComponents({ name: 'ElRadioButton' })[0],
      'is-active'
    )
  })

  describe('event', () => {
    it('change', async () => {
      const modelValue = ref('')
      const changeValue = ref('')
      const wrapper = mount(RadioGroup, {
        props: {
          modelValue,
          'onUpdate:modelValue'(v) {
            modelValue.value = v
          },
          onChange(v) {
            changeValue.value = v
          }
        },
        slots: {
          default: content.map((label) => h(RadioButton, { label }))
        }
      })
      await wrapper
        .findAllComponents({ name: 'ElRadioButton' })[1]
        .trigger('click')
      expect(modelValue.value).toStrictEqual('北京')
      await nextTick()
      expect(changeValue.value).toStrictEqual('北京')
    })
  })
})
