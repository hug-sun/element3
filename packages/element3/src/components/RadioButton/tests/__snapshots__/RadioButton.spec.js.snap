// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`radioButton.vue snapshot 1`] = `
<label
  aria-checked="false"
  class="el-radio-button"
  role="radio"
  tabindex="0"
>
  <input
    class="el-radio-button__orig-radio"
    tabindex="-1"
    type="radio"
    value="false"
  />
  <span
    class="el-radio-button__inner"
  >
    
    false
    
  </span>
</label>
`;
